### ISSUE TITLE
Superclass mismatch for class when updating measure

### ISSUE THREAD

**Issue by macumber** [NREL/OpenStudio#2658]
@rpg777 and @jmarrec have reported getting the error message 'Superclass mismatch for class' when attempting to update a measure.  Unclear if this is related to using old Ruleset::ModelUserScript base class or not.  Error can be reproduced by previously defining the class name in the same file.  Possible that error occurs when another measure exists with same class name?

    class SwapConstructions
    end

    # start the measure
    #class SwapConstructions < OpenStudio::Ruleset::ModelUserScript
    class SwapConstructions < OpenStudio::Measure::ModelMeasure

Related to https://github.com/NREL/OpenStudio/issues/507

### Rename your Measure
1. Use the "x2" button in the OpenStudio App's measure dialog to create a copy
2. Change the name from "MyMeasureName" to "SomeOtherName"
3. Delete the old copy of your Measure from your My Measures directory
### Make the superclasses match
1. Shut down the OS App or PAT
2. In measure.rb, change:
`class MyMeasureName < OpenStudio::Ruleset::ModelUserScript`
to:
`class MyMeasureName < OpenStudio::Measure::ModelMeasure`
or vice-versa, depending on what's in your measure.rb file.
3. Reopen the OS App or PAT

**Comment by macumber** [comment]
Prevents users from being able to update and use their measures so labeled as major

**Comment by macumber** [comment]
Possible that one idea is to load each measure into a unique module to prevent name collisions

**Comment by jmarrec** [comment]
I don't think this has to do with the old ModelUserScript or the new ModelMeasure, nor about class Name conflict. 

To me it seems it always comes from a measure or modeler description problem.

I ran into this error today again. I had something in my modeler description wiht special chars: `Watts @ 80/60/20`. Removing this fixed it. Weirdly enough, putting it back doesn't make `openstudio measure -u .` crash anymore...

**Comment by macumber** [comment]
I don't think this is specific to the modeler description, I think the key is that changing the measure.rb file results in a new file checksum and causes measure manager to recompute the xml file contents.  I have seen this clear the errors when modifying whitespace only (e.g. add a space somewhere).  The very weird thing to me is that the measure update check errors are intermittent.

**Comment by asparke2** [comment]
## Explanation of what's happening
Finally got to the bottom of this.  As @macumber guessed, it is the result of having 2 Measures with the same class name but derived from different superclasses both being updated by Measure Manager at the same time.  For me, this occurred when:
1. I had 2 copies of the Create DOE Prototype Measure:
    1. One in my local BCL (C:/users/aparker/BCL) 
    2. One in My Measures directory, which was pointing to my OpenStudio-measures github checkout
2. I had updated the superclass in my github copy from the (now deprecated) `OpenStudio::Ruleset::ModelUserScript` to the new `OpenStudio::Ruleset::ModelUserScript`

Here's an example ruby script that replicates the issue:

```ruby
# The old class (deprecated)
class UserScriptMeasure
end

# The new class that replaced it
class ModelMeasure
end

# Copy in BCL folder, derives from old class
class MyMeasure < UserScriptMeasure
end

# Copy in OpenStudio-measure github repo, derives from new class
class MyMeasure < ModelMeasure
end

### Run this file with ruby ###
# Result:
# example.rb:18:in `<main>': superclass mismatch for class MyMeasure (TypeError)
```

Basically, as the classes are being loaded, it first loads the class `MyMeasure`, which derives from `UserScriptMeasure`.  Then, it tries to load the second declaration of `MyMeasure`, which is derived from `ModelMeasure`.  Ruby doesn't know that `ModelMeasure` is just the renamed version of `UserScriptMeasure`, so it throws an error.

This is probably most likely to happen to developers who have both a local BCL and GitHub checkouts of the OpenStudio-measures.  However, it _can_ happen to any developer who happens to use a class name that was already used by another measure, but happens to derive from the opposite (deprecated/not deprecated) superclass.

**Comment by asparke2** [comment]
@macumber I downgraded to Normal since there is a workaround.  As you suggested, I think the real solution would be for Measure Manager to load each measure separately, unloading the previously loaded Measure before moving on.  I don't know what happens in PAT or the CLI if you have multiple measures that have different classes.

**Comment by asparke2** [comment]
### Update
The error can also happen if you edit the superclass in the measure.rb and then update the measure.xml files without closing Measure Manager first (because it has the previously-loaded class with the old superclass in memory still).  The solution is to shut down Measure Manager (the OS App) while editing measure.rb.

**Comment by DavidGoldwasser** [comment]
How do you address this issue if it happens on OpenStudio server? I have an empty "My Measures" directory and "BCL directory"

**Comment by DavidGoldwasser** [comment]
My solution, that seems to work is to change the offending measures from "OpenStudio::Measure::ModelMeasure" to "OpenStudio::Ruleset::ModelUserScript"

**Comment by jmarrec** [comment]
I can't draw any specific action items for this and it seems this is an issue than doesn't occur often, and only in specific develop environments. I'm going to close it, please reopen if you have a valid path forward.
